kale/embed/gripnet.py

```diff
@@ -241,3 +245,312 @@ def forward(
+            )
+
+    def forward(
+        self,
+        x: torch.Tensor,
+        edge_index: torch.Tensor,
+        edge_type: torch.Tensor = None,
+        range_list: torch.Tensor = None,
+        edge_weight: torch.Tensor = None,
+    ) -> torch.Tensor:
+        r"""
+        Args:
+            x (torch.Tensor): the input node feature embedding.
+            edge_index (torch.Tensor): edge index in COO format with shape [2, #edges].
+            edge_type (torch.Tensor, optional): one-dimensional relation type for each edge, indexed from 0.
+                Defaults to None.
+            range_list (torch.Tensor, optional): The index range list of each edge type with shape [num_types, 2].
+                Defaults to None.
+            edge_weight (torch.Tensor, optional): one-dimensional weight for each edge. Defaults to None.
+
+        Note: The internal feature layer is computed in the `forward` function of GripNet class. If the supervertex
+        is not a start supervertex, `x` should be the sum or concat of the outputs of the internal feature
+        layer and all external aggregation layers.
+        """
+
+        if self.setting.concat_output:
+            tmp = []
+            tmp.append(x)
+
+        if self.if_multirelational:
+            if edge_type is None or range_list is None:
+                error_msg = "`edge_type` and `range_list` are not set."
+                logging.error(error_msg)
+                raise ValueError(error_msg)
+
+        # internal feature aggregation layers
+        for net in self.inter_agg_layers[:-1]:
+            x = (
+                net(x, edge_index, edge_type, range_list)
+                if self.if_multirelational
+                else net(x, edge_index, edge_weight)
+            )
+            x = F.relu(x, inplace=True)
+            if self.setting.concat_output:
+                tmp.append(x)
+
+        x = (
+            self.inter_agg_layers[-1](x, edge_index, edge_type, range_list)
+            if self.if_multirelational
+            else self.inter_agg_layers[-1](x, edge_index, edge_weight)
+        )
+
+        x = F.relu(x, inplace=True)
+        if self.setting.concat_output:
+            tmp.append(x)
+            x = torch.cat(tmp, dim=1)
+
+        return x
+
+    def __repr__(self):
+        tmp = [f"\n    ({i}): {l}" for i, l in enumerate(self.inter_agg_layers)]
+
+        return "{}: ModuleList(\n  (0): InternalFeatureLayer: Embedding({}, {})\n  (1): InternalFeatureAggerationModule: ModuleList({}\n  )\n)".format(
+            self.__class__.__name__, self.in_channels, self.setting.inter_feat_channels, "".join(tmp)
+        )
+
+
+class GripNetExternalModule(Module):
+    """The internal module of a supervertex, which is an external feature layer.
+
+    Args:
+        in_channels (int): Size of each input sample. In GripNet, it shold be the dimension of the output embedding of the
```

Comment:
shold --> should
Did you use a spell checker in vscode? It should detect such typos.

Reply:
Thx, I just installed one.